docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+maintain [four separate connections][abci-conn]: One for "consensus" operations
+(BeginBlock, EndBlock, DeliverTx, Commit), one for "mempool" operations, one
+for "query" operations, and one for "snapshot" (state synchronization) operations.
+The rationale seems to have been that these groups of operations should be able
+to proceed concurrently with each other. In practice, it results in a very complex
+state management problem to coordinate state updates between the separate streams.
+While application authors in Go are mostly insulated from that complexity by the
+Cosmos SDK, the plumbing to maintain those separate streams is complicated, hard
+to understand, and we suspect it contains concurrency bugs and/or lock contention
+issues affecting performance that are subtle and difficult to pin down.
+
+Even without changing the semantics of any ABCI operations, this code could be
+made smaller and easier to debug by separating the management of concurrency
+and locking from the IPC transport: If all requests and responses are routed
+through one connection, the server can explicitly maintain priority queues for
+requests and responses, and make less-conservative decisions about when locks
+are (or aren't) required to synchronize state access. With independent queues,
+the server must lock conservatively, and no optimistic scheduling is practical.
+
+This would be a tedious implementation change, but should be achievable without
+breaking any of the existing interfaces. More importantly, it could potentially
+address a lot of difficult concurrency and performance problems we currently
+see anecdotally but have difficultly isolating because of how intertwined these
+separate message streams are at runtime.
+
+TODO: Impact of ABCI++ for this topic?
+
+### Context: RPC Issues
+
+The RPC system serves several masters, and has a complex surface area. I
+believe there are some improvements that can be exposed by separating some of
+these concerns.
+
+For light clients, some work is already underway to move toward using P2P
+message passing rather than RPC to synchronize light client state with the rest
+of the network.  Because of the way light client synchronization interacts with
+event logging, this is not as simple as just swapping out the transport: More
+details of the event system will need to be made visible via P2P, and that in
+turn has some implications for fault tolerance (in particular, I think the
+ability of a badly-behaved client to DDoS the network without punishment is a
+concern). However, once this work is done, it should in principle be possible
+to remove the dependency of light clients on the RPC service.
+
+For bootstrapping and operations, the RPC mechanism still makes sense, but can
+be further simplified. Here are some specific proposals:
+
+- Remove the HTTP GET interface entirely.
+- Simplify JSON-RPC plumbing to remove unnecessary reflection and wrapping.
```

Comment:
Is full compatibility with the 2.0 spec a target we want to aim for?

Reply:
Good question. My intuition is that isn't super important, though it would be nice to have. We don't rely on the missing bits anyway (e.g., notifications, batches).  If we decide to keep using JSON-RPC, one thing I might propose later (ADR) is to pull in a library implementation rather than our custom one (I wrote such a library a while ago). 

I don't think compatibility is as important as cleaning up some of the RPC dispatch plumbing we have now, though.